reconcile/external_resources/state.py

```diff
@@ -0,0 +1,131 @@
+    key: ExternalResourceKey
+    ts: datetime
+    status: ResourceStatus
+    resource_hash: Optional[str] = ""
+    reconcile_data: Optional[ExternalResourceReconcileData] = None
+
+
+class EnhancedJsonEncoder(JSONEncoder):
+    def default(self, obj: Any) -> Any:
+        if isinstance(obj, datetime):
+            return obj.isoformat()
+        elif dataclasses.is_dataclass(obj):
+            return dataclasses.asdict(obj)
+
+
+class ExternalResourcesStateManager:
+    def __init__(self, state: State, index_file_key: str):
+        self.state = state
+        self.index_file_key = index_file_key
+
+        try:
+            data = self.state[self.index_file_key]
+            states_list = parse_obj_as(list[ExternalResourceState], data)
+            self.index: dict[ExternalResourceKey, ExternalResourceState] = {
+                item.key: item for item in states_list
+            }
+        except Exception:
+            logging.info("No state file, creating a new one.")
+            self.index = {}
+
+    def _write_index_file(
+        self,
+    ) -> None:
+        data = [
+            item.dict(exclude=set(["resource_state"])) for item in self.index.values()
+        ]
+        self.state[self.index_file_key] = data
+
+    def _write_external_resource_states(
+        self,
+    ) -> None:
+        # Only need to save states that have been modified.
+        # Those that are present in the index state objects
+        resource_states = [
+            s.reconcile_data for s in self.index.values() if s.reconcile_data
+        ]
+        for state in resource_states:
+            key = f"{state.key.state_path}/qr_state.json"
+            self.state[key] = state.dict()
+
+    def read_resource_reconcile_data(
+        self, key: ExternalResourceKey
+    ) -> ExternalResourceReconcileData:
+        state_key = f"{key.state_path}/qr_state.json"
+        data = self.state[state_key]
+        resource_state = ExternalResourceReconcileData.parse_obj(data)
+        self.index[key].reconcile_data = resource_state
+        return resource_state
+
+    def get_external_resource_state(
+        self, key: ExternalResourceKey
+    ) -> ExternalResourceState:
+        obj = self.index.get(
+            key,
+            ExternalResourceState(
+                key=key,
+                ts=datetime.now(timezone.utc),
+                status=ResourceStatus.NOT_EXISTS,
+            ),
+        )
+        return obj
+
+    def set_external_resource_state(
+        self,
+        key: ExternalResourceKey,
+        state: ExternalResourceState,
+    ) -> None:
+        self.index[key] = state
+
+    def del_external_resource_state(self, key: ExternalResourceKey) -> None:
+        del self.index[key]
+
+    def get_all_resource_keys(self) -> list[ExternalResourceKey]:
+        return list[ExternalResourceKey](self.index.keys())
+
+    def save_state(self) -> None:
+        self._write_index_file()
+        self._write_external_resource_states()
```

Comment:
What's the relationship between these 2 files? Now I can see some transaction issues if one success the other fail.

Reply:
Each index item has an external_resource_state (ERS) stored in a separate file. The attribute has a value if the item has been reconciled.
write_external_resource_states iterates over all elements in the index and stores existent external_resource_states in their separate file. 

I chose this way to reduce the index file size; if no reconciliation is needed, there is no need to get the external_resource_state data.  The external_resource_state data is only used if a resource needs to be deleted (to get the Job image and input in case the spec is not present anymore).

So the only possible issue I foresee if the second call fails is that the deletion might get a different input data.... Let me sleep a bit on this to think about a better solution.

Reply:
So index file `external_resources_index.json` content is a list of all `ExternalResourceState` without `resource_state` field (where is this field defined?).

Both `ExternalResourceReconcileData` and `ExternalResourceState` have a same key field, do you want to use a one-to-one mapping between them and store in 2 different objects?

Let me try to grab the requirement here:

* Have a way to query all `ExternalResourceState`
* Have a way to query one `ExternalResourceReconcileData`

Put all `ExternalResourceState` in one index file may lead to concurrent issues, other storage such as dynamodb can use a different schema and index to better achieve it.

If we stick to S3 and use it as pure kv store, can use object path as a way embed info and to do query, so we can keep `ExternalResourceState` and `ExternalResourceReconcileData` in the same object, use list objects with path filter like `in-progress/key`, `success/key`. Just status change requires putting a new object in a different path and delete the old one.

Reply:
Oh sorry, `resource_state === reconcile_data`.  The reconcile_data is already part of the object but is stored in a separate file. 

```
class ExternalResourceState(BaseModel):
    key: ExternalResourceKey
    ts: datetime
    status: ResourceStatus
    resource_hash: Optional[str] = ""
    reconcile_data: Optional[ExternalResourceReconcileData] = None 
```
There is other data in de index entries like the object_hash, used to determine if a resource has changed. I fell like this is an early optimization. Maybe is a better idea (for now) to use one file per resources and lets make the optimization later.

Reply:
This changed a lot. PTAL again.